### PR TITLE
[internal/comparetest] Do not sort any slices by default

### LINKED ISSUE #17551: [internal/comparetest] Do not sort slices by default, use options instead
As a result of the discussion in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17286, it was decided not to sort all the slices by default but do it only if users ask for that explicitly with options like `IgnoreResourceMetricsOrder`, `IgnoreScopeMetricsOrder`

### PR DESCRIPTION
**Pull request by dmitryax**
Introduce options to ignore the order of log records, spans, metrics, and data points that should be used explicitly instead of sorting all slices by default.

This change includes reordering the expected metrics in scraper receivers since the metrics builder always emitts the metric slices sorted by metric names. The order of datapoints is defined by the order of applied `Record...Datapoint` functions which is sometimes unpredictable, so `IgnoreMetricDatapointsOrder()` option is applied in those cases.

Follow-up PR after https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17836

Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17551

### CODE REVIEW

File: internal/comparetest/logs.go [hunk 1]
```diff
@@ -132,34 +124,32 @@ func CompareResourceLogs(expected, actual plog.ResourceLogs) error {
...
-			alr := act.At(a)
+			alr := actual.At(a)
 			if _, ok := matchingLogRecords[alr]; ok {
 				continue
 			}
 			if reflect.DeepEqual(elr.Attributes().AsRaw(), alr.Attributes().AsRaw()) {
 				foundMatch = true
 				matchingLogRecords[alr] = elr
+				if e != a && outOfOrderErr == nil {
+					outOfOrderErr = fmt.Errorf("LogRecord with attributes %v expected at index %d, "+
```
Comment (djaglowski) [inline internal/comparetest/logs.go:150]: Is this type of error deduplicated because it may be too noisy?

File: internal/comparetest/util.go [hunk 2]
```diff
@@ -47,88 +47,282 @@ func getDataPointSlice(metric pmetric.Metric) pmetric.NumberDataPointSlice {
...
+	})
+}
+
+func sortMetricsInResourceMetricsSlice(rms pmetric.ResourceMetricsSlice) {
+	for i := 0; i < rms.Len(); i++ {
+		for j := 0; j < rms.At(i).ScopeMetrics().Len(); j++ {
+			rms.At(i).ScopeMetrics().At(j).Metrics().Sort(func(a, b pmetric.Metric) bool {
+				return a.Name() < b.Name()
+			})
+		}
```
Comment (djaglowski) [inline internal/comparetest/util.go:79]: Seems like these should both start with `ResourceMetricSlice` and handle the drilling/looping, or both should operate directly on the slice they are sorting.

**Review by djaglowski — APPROVED** [review]

File: internal/comparetest/logs.go (hunk 1 above)
Comment (dmitryax) [reply internal/comparetest/logs.go:150]: Yes, but we can show them all if you think it's a better approach

Comment (djaglowski) [reply internal/comparetest/logs.go:150]: I'm fine with it this way, just wanted to make sure I understood the motivation

Comment (dmitryax) [reply internal/comparetest/logs.go:150]: Maybe for consistency with extra/missing errors it's better to show all

File: internal/comparetest/util.go (hunk 2 above)
Comment (dmitryax) [reply internal/comparetest/util.go]: I addressed this comment and simplified name of all the helpers

**Pull request merged**